node/rate_limiter.js

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var assert = require('assert');
+
+var DEFAULT_SERVICE_RPS_LIMIT = 100;
+var DEFAULT_TOTAL_RPS_LIMIT = 1000;
+var DEFAULT_BUCKET_NUMBER = 20;
+
+function RateLimiterCounter(options) {
+    if (!(this instanceof RateLimiterCounter)) {
+        return new RateLimiterCounter(options);
+    }
+
+    var self = this;
+    self.index = 0;
+    self.rps = 0;
+    self.numOfBuckets = options.numOfBuckets;
+    self.buckets = [];
+    self.buckets.length = self.numOfBuckets;
```

Comment:
This is unnecessary turns the buckets array instance into a sparse array. I have seen claims that this turns the array into a hashtable internally, which probably performs poorly in comparison.

- http://www.html5rocks.com/en/tutorials/speed/v8/
- http://thibaultlaurens.github.io/javascript/2013/04/29/how-the-v8-engine-works/